src/expr/src/relation/canonicalize.rs

```diff
@@ -57,11 +117,29 @@ pub fn canonicalize_equivalences(equivalences: &mut Vec<Vec<MirScalarExpr>>) {
     equivalences.sort();
 }
 
+fn rank_complexity(expr: &MirScalarExpr) -> usize {
```

Comment:
This wants a doc comment.

Reply:
A nit, but why "non-leaves" instead of "nodes" which would simplify the logic a fair bit?

Reply:
This question still stands? If we want to prevent cycling, why not just "nodes"?

Reply:
It is true that "nodes" will also prevent cycling because an expression will only be able to be replaced with an expression with equal or fewer nodes. In my mind, having fewer function calls makes an expression "less complex," but ideally in the future, we could rank complexity by the estimated amount of time it takes to compute the expression.

Reply:
I believe it is true. The main property you want is that there is some measurement on the thing that must strictly increase when one thing is contained in the other, and strictly decrease when a greater thing is replaced by a strictly smaller thing. We can keep it as whatever, but "nodes" seemed much easier and if I swoop in and change it in the future is that a bug (is a thing I'd love the code to reflect).